src/main/java/org/wildfly/security/credential/source/OAuth2CredentialSource.java

```diff
@@ -0,0 +1,192 @@
+import org.wildfly.security.auth.client.AuthenticationContext;
+import org.wildfly.security.auth.client.AuthenticationContextConfigurationClient;
+import org.wildfly.security.credential.BearerTokenCredential;
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.util.HttpClient;
+import org.wildfly.security.util.HttpClient.HttpRequest;
+
+/**
+ * A {@link CredentialSource} capable of authenticating against a OAuth2 compliant authorization server and obtaining
+ * access tokens in form of a {@link BearerTokenCredential}.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class OAuth2CredentialSource implements CredentialSource {
+
+    /**
+     * Creates a new {@link Builder} instance in order to configure and build a {@link OAuth2CredentialSource}.
+     *
+     * @param tokenEndpointUri the token endpoint that will be used to obtain OAuth2 access tokens
+     * @return a new builder instance
+     */
+    public static Builder builder(URI tokenEndpointUri) {
+        return new Builder(tokenEndpointUri);
+    }
+
+    private final String grantType;
+    private final URI tokenEndpointUri;
+    private final Consumer<HttpRequest> clientAuthenticator;
+    private String scopes;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param grantType the OAuth2 grant type as defined by the specification
+     * @param tokenEndpointUri the OAuth2 Token Endpoint {@link URI}
+     * @param clientAuthenticator a consumer that will be called before sending an authorization request in order to setup client authentication
+     * @param scopes a string with the scope of the access request
+     */
+    private OAuth2CredentialSource(String grantType, URI tokenEndpointUri, Consumer<HttpRequest> clientAuthenticator, String scopes) {
+        this.grantType = checkNotNullParam("grantType", grantType);
+        this.tokenEndpointUri = checkNotNullParam("tokenEndpointUri", tokenEndpointUri);
+        this.clientAuthenticator = checkNotNullParam("clientAuthenticator", clientAuthenticator);
+        this.scopes = scopes;
+    }
+
+    @Override
+    public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+        return getCredential(credentialType, algorithmName, parameterSpec) != null ? SupportLevel.SUPPORTED : SupportLevel.UNSUPPORTED;
+    }
+
+    @Override
+    public <C extends Credential> C getCredential(Class<C> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+        if (BearerTokenCredential.class.equals(credentialType)) {
+            try {
+                HttpRequest request = HttpClient.builder().sslContext(resolveSSLContext()).build()
+                        .post(tokenEndpointUri.toURL())
+                        .param("grant_type", grantType);
+
+                clientAuthenticator.accept(request);
```

Comment:
I still don't like that we have this one-off HTTP request framework for this.  Right now we don't have a general HTTP client auth mech solution - but we do have an Authenticator which can get installed which uses the client configuration already.  Maybe we should just rely on that until we have client mechanism support.